### ISSUE TITLE
Extraction doesn't preserve empty types

### ISSUE THREAD

**Issue by cpitclaudel** [rocq-prover/rocq#12124]
#### Description of the problem

Extracting the following generates broken OCaml:

```coq
Module Type ty.
  Axiom t : Type.
End ty.

Module x <: ty.
  Inductive t' := .
  Definition t := t'.
End x.

Module z (inst: ty).
  Inductive xyz :=
  | XYZ (a: inst.t).
End z.

Module zx := z x.

Require Import Extraction.
Extraction zx.
```

This is because the `t` type is eliminated at extraction.  Is there a workaround?

#### Coq Version

8.10.2

**Comment by cpitclaudel** [comment]
Seems to be a duplicate of https://github.com/coq/coq/issues/5795

**Comment by herbelin** [comment]
@pi8027: Did you discuss about this issue with @letouzey?

**Comment by herbelin** [comment]
A possible workaround (communicated to me by @letouzey) is to bypass functor extraction in OCaml, the same way as it is already done in Haskell. The relevant line is the line `when lang () != Ocaml || Table.is_extrcompute ()` in `extract_env.ml` which could then be protected by a global option.

**Comment by herbelin** [comment]
The corresponding functor application would then be expanded into:
```
type __ = Obj.t
let __ = let rec f _ = Obj.repr f in Obj.repr f

module Coq_zx =
 struct
  type xyz =
  | XYZ

  (** val xyz_rect : (__ -> 'a1) -> xyz -> 'a1 **)

  let xyz_rect f _ =
    f __

  (** val xyz_rec : (__ -> 'a1) -> xyz -> 'a1 **)

  let xyz_rec f _ =
    f __
 end
```

**Comment by ejgallego** [comment]
@herbelin maybe an attribute could indeed work for this?

**Comment by herbelin** [comment]
> @herbelin maybe an attribute could indeed work for this?

Why not? @cpitclaudel would it be an convenient interface for that?

**Comment by cpitclaudel** [comment]
Sounds reasonable. But where should/would the attribute be applied? On all functors?
I guess the idea of functor extraction is that extracted Coq functors can be exposed as a richer API to OCaml code, when linking extracted code with handwritten code?
I tend to extract everything as one large flat file, so I wouldn't be bothered if it was just one option instead.

**Comment by pi8027** [comment]
> @pi8027: Did you discuss about this issue with @letouzey?

Yes. This is a typical module typing issue in extraction as in #10367, #5795, #5274, and #1957. (I think this issue is exactly a duplicate of #5795.)  Also, I have several examples of these issues with different flavors here: https://gist.github.com/pi8027/1a6050e5a90eeae1f4bb4de177d8ad08#file-1957-v.

IMO, we should implement an extraction flag to erase all the modules and functors from code properly. This is done in Haskell/Scheme extraction but it is broken and incomplete as in #8152, and also it does not support functors. Erasing modules and functors is exponential in general, but I hope this is not the case practically.
